### ISSUE TITLE
Nuget v2.0.2 - attestation option "direct" - TPM class

### ISSUE THREAD

**Issue by s21-himesh** [passwordless-lib/fido2-net-lib#305]
Hi,

We have noticed an issue.

When using the attestation option "direct" then clientDataJson appears to have 118 bytes when the tpm/ pubarea class expects 84. This throws an exception "throw new Fido2VerificationException("Leftover bytes decoding pubArea");" in TPM.cs line 634.

We were able to replicate the issue on your codebase changing attestation_type variable in custom.register.js to "direct".

We've temporarily changed our codebase to use "none" until fixed. Is there an alternative?

Please let me know if you need further details.

Thanks
Himesh

**Comment by aseigler** [comment]
I don't recall seeing anything like this before, do you have a sample clientDataJson that exhibits this behavior?

**Comment by s21-himesh** [comment]
Hi Aseigler,


Sorry for the late response. So decode of array below:-

'{"type":"webauthn.create","challenge":"ztuB2Bt0h_t4sPpMraDctA","origin":"https://localhost:41111","crossOrigin":false,"other_keys_can_be_added_here":"do not compare clientDataJSON against a template. See https://goo.gl/yabPex"}'

Thanks
Himesh

**Comment by aseigler** [comment]
I suspect this is going to be due to TPM attestations quietly being made with ES256 instead of RS256, which got fixed [here](https://github.com/passwordless-lib/fido2-net-lib/commit/b6f76420ba74f2a6a0486816f2cf36094b12fe9d).  We need to get a new build cut to get this on Nuget, I don't think we have one with this in it.

**Comment by m-gug** [comment]
No, the last release (3.0) doesn't have the changes as far as I can see. Would be nice to get the fix out.

**Comment by abergs** [comment]
@aseigler @m-gug 3.0.1 is available :shipit: 

https://www.nuget.org/packages/Fido2/3.0.1

https://github.com/passwordless-lib/fido2-net-lib/releases/tag/v3.0.1

**Comment by aseigler** [comment]
Please let us know ASAP if v3.0.1 does not fix the problem.

**Comment by m-gug** [comment]
Thanks for the quick release, the problem is solved, the registration is working again. :raised_hands:

However, another problem has emerged that did not exist before. On .Net 6, the System.Json.Text serializer now seems to choke on the return value in the MakeAssertion function.

It throws this exception:
"Serialization and deserialization of 'System.IntPtr' instances are not supported. Path: $.Result.AttestationCertificate.Handle."

The problem can also be recreated in your current master branch if you register a new credential in the demo app under Custom.

As a workaround I am not returning the complete result of MakeNewCredentialAsync(..), but returning a new CredentialMakeResult only with status="ok".
The frontend doesn't need the complete result, as far as I understand it, does it?

**Comment by abergs** [comment]
Thanks for the report!

> The frontend doesn't need the complete result, as far as I understand it, does it?
@m-gug The frontend is not interested in the result once the credential is created, only the CredentialCreateOptions in the first step of the dance.

**Comment by danjmaxcoil** [comment]
Hi, is there any solution about this bug?
We want to integrate Fido to our application and we need to know the authentication type the user authenticated with (fingerprint, face recognition, pin code...) so it is blocking us for going forward
Thanks

**Comment by abergs** [comment]
@danjmaxcoil Have you tried and used the 4.0 betas?
